core/src/main/scala/caliban/GraphQL.scala

```diff
@@ -95,68 +95,81 @@ trait GraphQL[-R] { self =>
-                                                                                config.validations
-                                                                              )
-                                                              _            <- checkHttpMethod(config)(executionReq)
-                                                            } yield executionReq
-                  executionRequest                     <- wrap(validate)(validationWrappers, doc)
-                  op                                    = executionRequest.operationType match {
-                                                            case OperationType.Query        => schemaToExecute.query
-                                                            case OperationType.Mutation     => schemaToExecute.mutation.getOrElse(schemaToExecute.query)
-                                                            case OperationType.Subscription =>
-                                                              schemaToExecute.subscription.getOrElse(schemaToExecute.query)
-                                                          }
-                  execute                               = (req: ExecutionRequest) =>
-                                                            for {
-                                                              queryExecution <- Configurator.configuration.map(_.queryExecution)
-                                                              res            <- Executor.executeRequest(req, op.plan, fieldWrappers, queryExecution, features)
-                                                            } yield res
-                  result                               <- wrap(execute)(executionWrappers, executionRequest)
-                } yield result).catchAll(Executor.fail)
+                  doc          <- wrap(Parser.parseQuery)(parsingWrappers, request.query.getOrElse(""))
+                  coercedVars  <- coerceVariables(doc, request.variables.getOrElse(Map.empty))
+                  executionReq <- wrap(validation(request.operationName, coercedVars))(validationWrappers, doc)
+                  result       <- wrap(execution(schemaToExecute(doc), fieldWrappers))(executionWrappers, executionReq)
+                } yield result).catchAll(Executor.fail(_))
               )(overallWrappers, request)
           }
+
+        private def coerceVariables(doc: Document, variables: Map[String, InputValue])(implicit
+          trace: Trace
+        ): IO[ValidationError, Map[String, InputValue]] =
+          Configurator.configuration.flatMap { config =>
+            if (doc.isIntrospection && !config.enableIntrospection)
+              ZIO.fail(CalibanError.ValidationError("Introspection is disabled", ""))
+            else
+              VariablesCoercer
+                .coerceVariablesEither(variables, doc, typeToValidate(doc), config.skipValidation)
+                .fold(ZIO.fail(_), ZIO.succeed(_))
```

Comment:
`ZIO.fromEither` is implemented as `ZIO.succeed(either).map(_.fold(ZIO.fail(_), ZIO.succeed(_))` which is 3 flatmaps instead of 1. It's a very micro-optimization, so I'm happy to revert it, but I thought since `VariablesCoercer.coerceVariablesEither` is not side-effecting might as well avoid the extra flatmaps